plugins/out_stackdriver/stackdriver.c

```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+        }
+
+        if (k.via.str.size == key_size && 
+            strncmp(key, (char *) k.via.str.ptr, k.via.str.size) == 0) {
+            /* make sure to free it after use */ 
+            ptr =  flb_strndup(v.via.str.ptr, v.via.str.size);
+            break;
+        }
+    }
+
+    return ptr;
+}
+
+/* 
+*    process_local_resource_id():
+*  - extract the value from "logging.googleapis.com/local_resource_id" field in the record
+*  - use extracted value to assign the label keys for differnet resource types that specified
+*    in the configuration of stackdriver_out plugin
+*/
+static int process_local_resource_id(const void *data, size_t bytes,
+                                     struct flb_stackdriver *ctx, char *type)
+{
+    int len_ptr;
+    int ret = -1;
+    size_t off = 0;
+    msgpack_object root;
+    msgpack_object_map map;
+    msgpack_unpacked result;
+    char *local_resource_id;
+    char *ptr;
+    const char delim[] = ".";
+
+    msgpack_unpacked_init(&result);
+    if (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        root = result.data;
+
+        if (!validate_msgpack_unpacked_data(root)) {
+            flb_plg_warn(ctx->ins, "unexpected record format");
+            return -1;
+        }
+
+        map = root.via.array.ptr[1].via.map;
+        local_resource_id = get_str_value_from_msgpack_map(map, LOCAL_RESOURCE_ID_KEY);
+        if (!local_resource_id) {
+            msgpack_unpacked_destroy(&result);
+            return -1;
+        }
+        
+        ptr = strtok(local_resource_id, delim);
+        len_ptr = strlen(ptr);
+        if (strncmp(type, "k8s_container", strlen("k8s_container")) == 0) {
+            /* check the prefix */
+            if (strncasecmp(ptr, type, len_ptr) != 0) {
+                free(local_resource_id);
+                msgpack_unpacked_destroy(&result);
+                return -1;
+            }
+
+            ptr = strtok(NULL, delim);
+            while (ptr != NULL)
+            {
+                /* Follow the order of fields in local_resource_id */
+                if (!ctx->namespace_name) {
+                    ctx->namespace_name = flb_sds_create(ptr);
+                }
+                else if (!ctx->pod_name) {
+                    ctx->pod_name = flb_sds_create(ptr);
+                } 
+                else if (!ctx->container_name) {
+                    ctx->container_name = flb_sds_create(ptr);
+                }
+
+                ptr = strtok(NULL, delim);
+            }
+
+            if (!ctx->namespace_name || !ctx->pod_name || !ctx->container_name) {
+                flb_free(local_resource_id);
+                msgpack_unpacked_destroy(&result);
+                flb_sds_destroy(ctx->namespace_name);
+                flb_sds_destroy(ctx->pod_name);
+                flb_sds_destroy(ctx->container_name);
+                return -1;
+            }
+        }
+        else if (strncmp(type, "k8s_node", strlen("k8s_node")) == 0) {
```

Comment:
don't use strlen as a nested value. If the string you are comparing has a fixed value, there is no need to calculate it over and over.

Reply:
updated